### ISSUE TITLE
difficult assignment of lyrics to correct note line

### ISSUE THREAD

**Issue by Bacchushlg** [Audiveris/audiveris#334]
![The Grass Grows Greener (2)](https://user-images.githubusercontent.com/33522630/70387035-6baa3500-19a0-11ea-88eb-0c8698141a34.jpg)

In some seldom cases Audiveris has problems to properly assign the lyrics to the correct note lines.
Here we have 2 problems:

- The system are so near together that the bottom lyrics line is assigned as lyrics on top of the 2nd system.
Maybe a query that is already provide in case of notation symbols (select line above/below) could help in such a case. Or, more sophisticated: try both cases and select the one with the best fit "sillable - note head".

- The 2nd concerns the text below the last lyrics line. Here we find a couple of text that is not lyrics line but assigned a such.
Also here a minimum fit "sillable - note head" (line per line from top) may be used to find a proper decision.

2nd

**Comment by hbitteur** [comment]
@Bacchushlg 
I'm afraid the problems with lyrics in this particular score can't be fixed by any reasonable algorithm. The reason is that the page is way too crowded.

Instead, I intend to work on your other proposal to provide better lyrics-oriented manual corrections (with voice, verse, etc).
At the end of the day, a few manual features may well overcome tricky and non-reliable algorithms...

**Comment by Bacchushlg** [comment]
[Que sera sera.zip](https://github.com/Audiveris/audiveris/files/4050467/Que.sera.sera.zip)

I add this score as additional example for difficult voice assignment.
I just tried to transcribe while activating the flag "lyrics even located above staff", but none of the lyrics above where assigned correctly

**Comment by hbitteur** [comment]
Yes, the new algorithm to dispatch inter-system lyric lines does not take into account the "lyric ABOVE staff" switch. Let me try to adapt the algorithm to this case, I need to study this more thoroughly.

It does not seem to be easy.
For example, let's have a look at the vertical interval between system 1 and system 2. In this interval, we have 3 lines:
1. "1. Doo ba doo.  When she was a"
2. "C   C   G7  G7  G"
3. "moth - er:  What will I be,    will I be"

We, as human beings much smarter than a dumb software, can infer that:
- Line 1 is made of lyrics and relates to upper system (1).
- Line 2 is made of chord names and related to lower system (2).
- Line 3 is made of lyrics and relates to lower system (2). But where are the words that precede "moth - er"? They should be above system 1 on its right side. But there are none there!

Also, how can the software make the distinction between the 3 lines above? We can play with the vertical gaps between lines and between lines and staff lines but this might be very touchy.

And a final note about the following case, between the last 2 systems of the page:

![image](https://user-images.githubusercontent.com/22053634/72224276-f3272d00-3578-11ea-9905-0fc1f43ec0c1.png)
Notice the mess just above the lower system:
- Chord names jump from one line to the other.
- The last line is a mix of: 16 me:   C7  mf SEGNO F", what can the program do? seriously?

**Comment by hbitteur** [comment]
OK, here is something that would work on this case and perhaps others (assuming the option "_lyrics even above staff_" is ON):
1. Instead of staff line, use the staff "border" which is the staff line vertically shifted to include timing symbols (chords, beams, articulations, ...). This is the real end of staff region. 

2. Measure vertical distances as follows:
   - between upper staff border and first text line
   - between first line and second line
   - ...
   - between second to last line and last line
   - between last line and lower staff border

Then pickup the largest vertical distance. This gives the "split gap" between which lines belong to upper system and which lines belong to lower system.

**Comment by Bacchushlg** [comment]
1. good idea!
2. Might work.
But to say it once more: I don't expect that Audiveris can detect all these cases. Instead I'd like to be able to assign the lyrics correctly.
This also applies to the cases where lyrics switch between 1st and 2nd voice of a note line.

The only automatic approach that I could imaging is:
- take a complete lyrics line
- try to assign to a voice of the note line (above, 1st time)
- calculate a trust value: sum of the distances between the middle of each syllable to it's assigned note head vertical position
- repeat for the other voices
- repeat for the note line below (if enough probable)

In case of multi-verse lines, it is enough to check the first line (the other's belong to the same voice)

**Comment by hbitteur** [comment]
I agree, however smart the program can be, the user should always have the last word. And for that we need to provide convenient ways to do that manually.

**Comment by hbitteur** [comment]
You said "calculate a trust value: sum of the distances between the middle of each syllable to it's assigned note head vertical position"
Does this mean that the voice whose note line is (in average) the closest to the lyrics first line will be assigned to these lyrics?

**Comment by Bacchushlg** [comment]
Exact

Am 13. Januar 2020 20:42:11 MEZ schrieb "Hervé Bitteur" <notifications@github.com>:

**Comment by hbitteur** [comment]
Just to make sure that I understand everything clearly, here is the beginning of 2nd system of "Que sera sera" score.

![image](https://user-images.githubusercontent.com/22053634/72637174-fba8aa80-3960-11ea-8cd0-9fcaf98d51e1.png)

This is not the raw output of OMR engine yet :-), I had to make several manual modifications to get to this result. But I'd like to check if you agree with the final mapping, as described below.

There are 2 parts in this system, the upper one S/A (Soprano/Alto) and the lower one T/B (Tenor/Bass).
In each part, there are two voices, voice 1 displayed in blue, voice 2 displayed in green.

We begin by a "line" of chord names (C, C, G7, G7)
Then, regarding the lyrics, we can see 4 lyric lines from top to bottom:
1. Line "moth - er: What will I be," relates to the blue notes below (A4 etc).
Part 1 (SA), voice 1, verse 1, above
2. Line "moth    -    er: What will I, what will I be?" relates to green notes above (F4 etc).
Part 1 (SA), voice 2, verse 1, below
3. Line "ooh, ----- ..." relates to the blue notes below (C4 etc).
Part 2 (TB), voice 1, verse 1, above
4. Line "girl.   ..." relates to the green notes above (C3 etc).
Part 2 (TB), voice 2, verse 1, below

Audiveris assigns unique voice numbers per part (not per system)
And in this score, there is just one verse!

Could you confirm all this?

**Comment by Bacchushlg** [comment]
That's it!

**Comment by hbitteur** [comment]
Hi Baruch,

Could you give me a precise definition of "verse number", or point to some existing documentation.

And how does it translate to MusicXML? For your information,  here is the  [lyric element](http://usermanuals.musicxml.com/MusicXML/MusicXML.htm#EL-MusicXML-lyric.htm%3FTocPath%3DMusicXML%2520Reference%7CScore%2520Schema%2520(XSD)%7CElements%7Cnote%7C_____66)  definition in MusicXML 3.0 specs. What they call "number" might be the verse number, but this is just a guess.

**Comment by hbitteur** [comment]
OK, here is an improved version via commit 5c541a50049ba3fd6fc2f6cd18ff71ee0afe62f0 in development branch.

Notice that OMR engine should now correctly dispatch lines to systems and staves.
If not, a simple manual link from a lyric item to its related head chord, provokes the switch of the whole lyric line.

You can see how this this new version behaves on scores like "Que sera sera",  "Lila Tilla", etc.
Doc is still pending, don't hesitate to ask if something is not clear.

Question for @Bacchushlg: I'm not sure of the meaning of verses in "Que sera sera". Are the various lines in each part real verses, or rather lyrics sung by say Soprano vs Alto or Tenor vs Bass. In "Lila Tlla", these lines definitely look like verses to me...
Please help me disambiguate these cases.

**Comment by Bacchushlg** [comment]
[Que sera sera.zip](https://github.com/Audiveris/audiveris/files/4129027/Que.sera.sera.zip)

It's almost perfect now!

I append the OMR file for "que sera" after my corrections. 
And yes: there are no verses - all different lines are different voices.

Some minor remarks: in page 2, measure 1, the lower note line has exchanged voices: voice 1 is the lower one. Therefore the lyrics are wrong-assigned, too.

The export of the voices (above/below note line) is not correct: they finally appear as verse 1 and verse 2 (at least they appear as such in MuseScore).

The lyrics lines above the note lines are not recognized as lyrics, although I had set the flag (support for lyrics even above...) in the score settings. Are you sure that you consider this flag?

**Comment by hbitteur** [comment]
Between "almost perfect" and "perfect" is where we find room for user manual correction! By definition :-)

Latest algorithm for linking lyric syllable and head chord is as follows:
1. [Target staff has been determined upfront for the whole lyric line],
2. Syllable reference point is computed (ordinate at text baseline, abscissa at 1/4 of syllable width),
3. A range in abscissa filters head chords candidates around the syllable,
4. The smallest euclidian distance between syllable reference point and chord area center gives the winner.
5. If another syllable of the same lyric line is already linked to this head chord, it's the closest in abscissa which wins. The discarded syllable is then given a second choice.

And the syllable voice is set to be the voice of the linked head chord.

Now, this does not guarantee that the head chord voice is correct. If needed, you can manually force head chord voice, this is done at chord level, not at syllable level.
You can:
- Force 2 chords to share the same voice,
- Force 2 chords to use separate voices,
- Force 2 chords to share the same time slot,
- Force 2 chords to use separate time slots.
To do this, just point to one chord and then to the other (pressing CTRL key), and open the popup menu (right click). If relevant, you'll find a submenu "chords" offering the relevant actions I listed above.

This is one example for which I should update the user documentation ASAP. I'm investigating the GitHub Pages approach. More on this later.

**Comment by hbitteur** [comment]
@Bacchushlg said: 

> And yes: there are no verses - all different lines are different voices.
>...
> The export of the voices (above/below note line) is not correct: they finally appear as verse 1 and verse 2 (at least they appear as such in MuseScore).

Baruch, I asked you this question some time ago: Could you point me to a good definition of what a "verse" is? 
And especially, how can we disambiguate between verses and voice lines?

According to your examples (Que sera sera, Lila Tilla), could we assume that:
- Lyric lines separated by a staff (typically line(s) above the staff and line(s) below the same staff) belong to separate voices (one voice for lines above, another voice for lines below).
- Lyric lines on the same side of the staff they belong to, they all belong to the same voice but are separate verses.

**Comment by hbitteur** [comment]
@Bacchushlg asked:
> The lyrics lines above the note lines are not recognized as lyrics, although I had set the flag (support for lyrics even above...) in the score settings. Are you sure that you consider this flag?

Yes, this flag is supported. Otherwise you could not see any lyrics recognized above the staff, at all.

This does not say that any line located above the staff is "blindly recognized" as a lyric line.
This applies only to the first line located just above the staff, within a certain maximum vertical distance from the staff "core border".

![image](https://user-images.githubusercontent.com/22053634/73435927-deed7900-4349-11ea-9091-bd3d45d23265.png)

The snippet above was taken from page 4 of "Que sera sera" score, between systems 3 and 4.  Notice the complexity in this amount of text lines.
The upper line "Que, - - que sera , etc.." was correctly recognized as lyric and assigned to the upper staff.
For the lower staff, raw results are not what we would expect. Perhaps because of the "85" word on the left side, perhaps considered as the lyric line above the staff. Perhaps because the line "be, ... Will-- be." lies too far above the staff core border (the line displayed in orange), because of the presence of crescendo sign (recognized in CURVES step, thus after the TEXTS step).

![image](https://user-images.githubusercontent.com/22053634/73435791-9a61dd80-4349-11ea-909e-31ca71ad34b7.png)

Same page 4 as before, system 2.
There is no way the OMR engine could detect that the beginning of the  first physical line located right above the staff is in fact composed of a few chord names (C and C) and then a lyric line portion, while chord names (C, G7, G7)  are now located in yet another line above this one.
The current algorithm requires that all words of a line match chord name pattern, not just half of it.

In these 2 cases, you as the end user must come to the rescue, and manually assign lyric portions and text portions. To do so, use a lasso and assign items using proper "lyric" or "text" button in the shape board.

**Comment by Bacchushlg** [comment]
Ah, I understand. So the initial measure number may completely disturb the further recognition.
Do you think that it could make sense to ignore the initial part above each system during text recognition. E.g. everything that is left of the right border of the clefs.

**Comment by Bacchushlg** [comment]
about voices and verses: from my point of view Audiveris did properly recognize the voices (I can see it from the colour of the lyrics). But the export is not correct: the lyrics of the 2nd voice is exported as 2nd verse. See the appended extract of the XML-file: 
lyric number="2" 
must be "1", because it is the 1st verse (of the 2nd voice).

```
      <note default-x="62">
        <pitch>
          <step>A</step>
          <octave>4</octave>
        </pitch>
        <duration>1</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem default-y="10">up</stem>
        <lyric number="1" default-y="24">
          <syllabic>begin</syllabic>
          <text>moth</text>
        </lyric>
      </note>

      <note default-x="61">
        <pitch>
          <step>F</step>
          <octave>4</octave>
        </pitch>
        <duration>1</duration>
        <voice>2</voice>
        <type>quarter</type>
        <stem default-y="-61">down</stem>
        <notations>
          <slur type="start" number="1" bezier-x="15" bezier-y="-13" default-x="6" default-y="-52" placement="below"></slur>
        </notations>
        <lyric number="2" default-y="-102">
          <syllabic>begin</syllabic>
          <text>moth</text>
        </lyric>
      </note>

```

**Comment by Bacchushlg** [comment]
I just notice that MuseScore gets problems with this... The lyrics line now overlap.
I just had a look at the Musicxml defintion.
There is an attribute handling the lyrics position: "placement" (above/below)
I just tried to add this to the xml file 

```
        <lyric number="1" default-y="-102" placement="above">
          <syllabic>begin</syllabic>
          <text>moth</text>
        </lyric>
```
but MuseScore seems to ignore this.
In principle, MuseScore support the lyrics position since version 3, but is seems that the import ignores this flag. Also in the export of MuseScore as xml I don't find this parameter. Export an re-import "forgets" the lyrics position...

**Comment by hbitteur** [comment]
@Bacchushlg back to your remark about "Que sera sera" score:
> Some minor remarks: in page 2, measure 1, the lower note line has exchanged voices: voice 1 is the lower one. Therefore the lyrics are wrong-assigned, too. (Same in page 4, 1st system, measure 4)

Here below is the raw output of OMR engine (for page 2, measure 1):
![2020-02-03 15_05_31-Que sera sera - Audiveris Music Scanner](https://user-images.githubusercontent.com/22053634/73660539-a7a80080-4698-11ea-90ee-9a71468952ce.png)
The lyric syllables are correctly linked to head chords, the problem is that head chord voice (and thus displayed color) is wrong.

The two chords indicated by red arrows should belong to the same voice.
Without digging into the (fragile) voice algorithm, I'm pretty sure that, seen from the quarter rest symbol, the lower C3 (with downward stem) got preferred to the upper A3 (with upward stem), simply because it is closer ordinatewise, and there is no other competing chord nearby.
Yes, it's the wrong choice, but what else could OMR do in this case?
The original song writer could have inserted quarter rests for each voice, but that's another story.

Anyway, the end-user can now come to the rescue.
For example, by selecting the 2 chords indicated by the red arrows and expliclty ask them to share the same voice, as depicted below:
![2020-02-03 15_11_39-Que sera sera - Audiveris Music Scanner](https://user-images.githubusercontent.com/22053634/73661376-408b4b80-469a-11ea-8faa-dd4909f68ea4.png)

Remark: You could instead select the quarter rest and the lower head chord (C3) and explicitly ask them to use separate voices. The result would be the same.

Either way, the final result is:
![2020-02-03 15_34_52-Que sera sera - Audiveris Music Scanner](https://user-images.githubusercontent.com/22053634/73661854-19814980-469b-11ea-8a76-4ea0276ff43d.png)

The other example (page 4, measure 4) derives from the same problem.

**Comment by hbitteur** [comment]
Back to comment https://github.com/Audiveris/audiveris/issues/334#issuecomment-580247130 about lyric numbers.

Remark: Right now, Audiveris does not export the placement attribute (above / below). I can add this if you wish.

Audiveris does not yet handle the notion of verse number, just the notion of lyric number within the same part. And the number is assigned sequentially from top to bottom within the same part.

In this score (que sera sera), most parts have both one line above and one line below the part single staff. They get numbered 1 and 2, respectively.
But in the very first system, the upper part exhibits just one lyric line (below the staff). it is numbered 1, because it's the first and only one in this system part.

There is yet no notion of verse in this numbering scheme. I have re-read MusicXML specification and I still can't figure out their precise notion of lyric "number" vs  lyric "name", especially because the only explanation I found there was:
> The lyric number indicates multiple lines, though a name can be used as well

Should we put the verse value into the "number" attribute? Since we have no verse in this page, but just separate voices, all numbers should set to 1...
What can we put into the lyric "name" attribute. The sole example in MusicXML spec uses the word "verse" as the content of "name" attribute. This does not help me a lot!!!

The only progress I could envision is to feed the "placement" attribute. Would this help? I don't know.

**Comment by Bacchushlg** [comment]
The lyrics "name" is not need. Just set the number as number of the verse, as you do it already now. 
The problem that I described above: you have set the lyric number to "2", too. But it is the 1st verse (of voice number 2), so the correct lyrics number must be 1.

```
        <voice>2</voice>
        <type>quarter</type>
        <stem default-y="-61">down</stem>
        <notations>
          <slur type="start" number="1" bezier-x="15" bezier-y="-13" default-x="6" default-y="-52" placement="below"></slur>
        </notations>
        <lyric number="2" default-y="-102">    <== this must be 1, not 2
          <syllabic>begin</syllabic>
          <text>moth</text>
        </lyric>
```

**Comment by hbitteur** [comment]
We have a misunderstanding here.

I (the OMR engine) don't know what a verse is, I just play with lyric lines, part by part. And, in a part, the lyric lines have different numbers, from top to bottom. They are not numbered within the same voice but within the same part.

If I understand your remark correctly, I should number the lines within all the lines that relate to the same voice (and I think that these lines are true verses).
In the case of this particular score, all lyric line numbers would be set to the same 1 value.

So, we would get (for any system) in our current example (Que sera sera)::
- Part 1:
    - above: voice 1, lyric above number 1
    - below, voice 2, lyric below number 1
- Part 2:
    - above: voice 1, lyric above number 1
    - below, voice 2, lyric below number 1

But in score "Lila Tilla", where lyrics are all below staff, we would have:
- Single staff:
    - above (nothing)
    - below, 4 lines, numbered from 1 to 4 (real verses).

To do so, perhaps we could consider that the staff is a kind of border. Above the staff we have at most 1 line of lyrics and below the staff we can have 1 or several lines, numbered accordingly 1..n

Would this make sense?

**Comment by hbitteur** [comment]
I just implemented this new lyrics numbering scheme, which uses staff relative placement.
Available in commit 4252a531c3ae0a7e26a1ca2f144be1bd87ee69e3 on development branch.

**Comment by Bacchushlg** [comment]
perfect now.
What is missing is that MuseScore handles the input correctly. I have running a demand for this since more than 1 year...

**Comment by hbitteur** [comment]
I just checked Finale on whole score "Que sera sera", with few user manual corrections. 
The lyrics lines and recognized chord names are well handled (I will push the latest commit on chord names in a few minutes).

It looks good and sounds good as well!
Believe me, this was a great moment of intense pride! :-) :-)

See below the first of the 4 pages in Finale:

![image](https://user-images.githubusercontent.com/22053634/73754448-b8be4380-4764-11ea-9f72-72c7ca8dc2b0.png)

**Comment by hbitteur** [comment]
Here is the commit b5611473d7801be395b198d96c4c24466cd81598 on development branch to support chord names.

**Comment by Bacchushlg** [comment]
Marvellous!!!
You really have a good reason to be pride!!!
What a big step to perfect music OCR!
I tend to re-transcribe a couple of my score due to the much better lyrics recognition plus chord  evaluation.

Just a little question: were you able to solve the wrong voice recognition in measure 14, 1st page inside Audiveris? I always get a 3rd voice in this measure and find no way how to fix.

**Comment by hbitteur** [comment]
I suppose you are referring to this case, where we can see 3 voices:
![image](https://user-images.githubusercontent.com/22053634/73764009-90d5dc80-4772-11ea-8a4d-15ba27442c68.png)

If so, then it's easy: Force the two lower left chords to share the same voice.
Click on first chord, then press CTRL and click on second chord.
Then a right click gives this menu:
![image](https://user-images.githubusercontent.com/22053634/73764385-283b2f80-4773-11ea-8d34-68917c172856.png)

Et voilà !
![image](https://user-images.githubusercontent.com/22053634/73764505-74866f80-4773-11ea-9347-b2760274921b.png)

**Comment by Bacchushlg** [comment]
so simple... ;-))

**Comment by hbitteur** [comment]
Baruch, 

In parallel with improvements and bug fixing, I'm preparing the documentation for coming 5.2, whose main features relate to end-user manual corrections.
I will start from the 5.1 handbook and update/augment it for 5.2 release.

I have already migrated GitBook files to Github Pages, this will be more convenient since Audiveris doc files will be hosted under the /docs folder of our github repository and directly available on audiveris.github.io page.
I still need to select a theme suitable for product documentation, there are several ones to pick from. I'll let you know when this "5.1 content" starting point is available.

Then I'll work on doc modification, especially to explain what the user can do and how.
And I'm thinking of this "Que sera sera" example, with exhibits typical situations for the end-user. Do you think I could use it (or part of it) as an underlying example on which the end-user could operate? Are there any copyrights on this score or image?

Please let me know.
Cheers,
/Hervé

**Comment by MusicForStringsPercussionCeleste** [comment]
"Que sera, sera" composer died in 2001 so the song and lyric are still under copyright, but if you use only a part of the song as part of an education project you can claim Fair Use safe harbor.  The worst that would happen is a DMCA take down notice.
thanks for your work on the documentation.
cheers,
Music for SPC


________________________________

**Comment by hbitteur** [comment]
@MusicForStringsPercussionCeleste 
Thanks for your remark.
In fact, we don't really need to publish a whole page, but just a few portions of the image in order to show a typical situation. Such a snippet generally describes something the OMR engine has failed to transcribe correctly and the documentation will explain how the end-user can manually fix this.
This is similar to the snippets we frequently post on this forum, and I don't consider we would infringe copyright in any way.

**Comment by weidi** [comment]
Hi, just trying to figure out a way on how to accomplish voice assignment here.
![image](https://user-images.githubusercontent.com/2188062/87469399-840bbd80-c61b-11ea-9047-cd86d5576312.png)
For me it seems that a seperate voice has to have it´s own stem but this is not really the case here.
Do you have any hint or advise?
Cheers weidi

**Comment by weidi** [comment]
Here is the manual changed lines with arrows pointing to the problematic chords.
You can see that i splitted chords and added Flags that timing is correct again. Using this to and export to musicxml finally gives me two voices for each Line, hope that makes it clearer?
![image](https://user-images.githubusercontent.com/2188062/87585172-f396b100-c6de-11ea-852f-afbf382b8e62.png)

**Comment by hbitteur** [comment]
OK, I see the point now.
The root problem lies in this construction (original image): 
![image](https://user-images.githubusercontent.com/22053634/87699387-1388ac00-c795-11ea-8178-54554f2e445b.png)
As compared with this (after your manual modification): 
![image](https://user-images.githubusercontent.com/22053634/87699586-5d719200-c795-11ea-8fd2-6f266e462977.png)

It boils down to a question of stem length. Even if you look very carefully at the original image (gray or black & white) you, as a human being, may not always be able to tell whether we have one long stem or two aligned but separate stems.
And the OMR engine is much less powerful than a human being...

If there is just one stem, then we have just one chord and one voice.
If there are two separate stems perfectly aligned vertically, then we have two separate chords, and their voices must be different because the two chords start at the same time slot.

However, since there is a beam that links the two upper notes, we can say that these stems are indeed long stems (otherwise we would have 8th notes aligned with quarter notes).
So, we are back to the first case: one stem, one chord, one voice.

If the author wanted to write separate voices, he would have used down stems for the lower two notes, and a beam at the bottom to link these two notes (kind of a mirror of the upper two notes).

We could also dream of a super-smart OMR able to dispatch notes heads stuck to the same stem to various voices. 
But then, imagine the situation (measure  number 4 of your own example): 
![image](https://user-images.githubusercontent.com/22053634/87701525-6c0d7880-c798-11ea-90f9-0306463c4fc3.png)
Along the right-most stem , we have 3 note heads. How should we dispatch each of these 3 heads? According to your manual modification, you've decided to assign the upper 2 heads to a voice and the 3rd one to another voice. But you could as well have assigned the upper head to a voice and the two lower heads to another voice. Why not?

In other words, this would open a Pandora box... And current algorithm for  time slots & voices is already tricky enough!

**Comment by weidi** [comment]
Thank you for your valuable feedback and description on whats happening and why!
The problem I see here is when not taking care manually i end up having 8 voices that need to be manually aligned in Musescore or  others. 

I just tried another "fix" for it by adding a manual rest in between.
![image](https://user-images.githubusercontent.com/2188062/87719116-2ad69200-c7b3-11ea-9f92-a8ca4b04dfa9.png)
This makes it again being displayed as a second voice but keeping the first voice intact.

I more and more see that this is a extremely complex scenario and not sure how to really circumvent such a scenario.
BTW: keep up the great work and enthusiasm, really enjoy it!
